### PR TITLE
[lexical][lexical-clipboard] Bug Fix: Preserve alignment when pasting selected text

### LINKED ISSUE #8101: Bug: Text alignment is lost on copy-paste
**Lexical version:** latest (tested on https://playground.lexical.dev)

## Steps To Reproduce
1. Create multiple paragraphs.
2. Set the **first paragraph** alignment to **center** or **right**.
3. Copy the first paragraph.
4. Move the cursor to another paragraph that has **default left alignment**.
5. Paste the copied content.

[Demo](https://youtu.be/f83vxQGogIs)

## Link to code example
https://playground.lexical.dev/

## The current behavior
The pasted text is reset to **left-aligned**, even though the copied text was center- or right-aligned.

## The expected behavior
The pasted text should **preserve its original alignment** (center or right).

## Impact of fix
This affects basic formatting workflows. Users expect copy-paste to retain alignment, especially in document-style editors.

### PR DESCRIPTION
**Pull request by Rohithmatham12**
## Description

When copying selected text from an aligned paragraph inside the same Lexical namespace, the `application/x-lexical-editor` clipboard payload took priority over `text/html`. That JSON payload only contained the selected `TextNode`, so pasting into another paragraph reused the destination paragraph's default alignment and discarded the source paragraph alignment.

This updates clipboard JSON export to retain the non-inline parent element when the selected child content covers the full element text. That preserves block-level formatting such as paragraph alignment for full-paragraph text selections while partial text selections remain text-only.

Closes #8101

### Before

Copying all visible text from an aligned paragraph and pasting it elsewhere lost the source paragraph alignment.

### After

Full-paragraph text selections preserve block-level alignment on paste, including paragraphs that contain inline elements such as links, while partial text selections remain text-only.

Automated tests:

- `node node_modules/vitest/vitest.mjs --project unit packages/lexical/src/__tests__/unit/HTMLCopyAndPaste.test.ts --run`
- `node node_modules/vitest/vitest.mjs --project unit packages/lexical-clipboard/src/__tests__/unit/GetClipboardDataExtension.test.ts packages/lexical-clipboard/src/__tests__/unit/ClipboardImportExtension.test.ts --run`
- `node node_modules/prettier/bin/prettier.cjs --check packages/lexical-clipboard/src/clipboard.ts packages/lexical/src/__tests__/unit/HTMLCopyAndPaste.test.ts`
- `node node_modules/eslint/bin/eslint.js packages/lexical-clipboard/src/clipboard.ts packages/lexical/src/__tests__/unit/HTMLCopyAndPaste.test.ts`

### CODE REVIEW

**Review by potatowagon — COMMENTED** [review]
Reviewed by Navi (Tater Thoughts Bobblehead) on behalf of @potatowagon.

**Assessment: LGTM with minor note** ✅

**What this does:** Fixes #8101 — when copying text from an aligned paragraph within the same Lexical editor, the `application/x-lexical-editor` clipboard JSON only contained the TextNode, losing block-level formatting (alignment, indent). The fix adds `$shouldIncludeElementWithSelectedChildren()` which includes the parent element in the clipboard payload when the selected children cover the full element text content.

**What I checked:**
1. **Logic correctness:** The function checks: (a) is range selection, (b) node is non-inline ElementNode, (c) not root, (d) selectedNodes content under this parent matches the parent's full text content. This correctly identifies "entire paragraph selected" without false positives on partial selections.
2. **Edge cases considered:**
   - Partial text selection → `selectedTextContent !== node.getTextContent()` → returns false ✅
   - Empty paragraph → `selectedTextContent !== ""` guard → returns false ✅
   - Inline elements (links, marks) → `node.isInline()` guard → skipped ✅
   - Root node → `$isRootNode(node)` guard → skipped ✅
   - Multiple children in paragraph (e.g. bold+plain) → uses `$sliceSelectedTextNodeContent` per node → correct ✅
3. **Performance note:** The function iterates selectedNodes and calls `isParentOf()` for each. For very large selections this is O(n×depth), but clipboard operations are user-initiated and typically small. Acceptable.
4. **Test coverage:** Good test — creates a centered paragraph, copies via `$getClipboardDataFromSelection`, pastes into another paragraph, asserts alignment is preserved in DOM output.
5. **www compat:** No API signature changes. `$shouldIncludeElementWithSelectedChildren` is module-private. The behavioral change (richer clipboard payload) is strictly additive — paste targets that ignore element wrappers will work as before.

**One observation:** The text-content comparison (`selectedTextContent === node.getTextContent()`) is a heuristic — it could theoretically match if a partial selection happens to produce the same string as the full element (e.g., repeated characters). This is an extremely unlikely edge case and the consequence (including alignment where it shouldn't) is benign.

**CI:** CLA ✅, Vercel ✅. Full test matrix pending (PR is ~30min old). Recommend waiting for unit green.

Ready to approve once full CI passes.

**Comment by etrepum** [comment]
Title and description doesn't match the [pull request template](https://raw.githubusercontent.com/facebook/lexical/refs/heads/main/.github/pull_request_template.md)

**Comment by mayrang** [comment]
The approach makes sense, but `$shouldIncludeElementWithSelectedChildren` breaks on paragraphs containing inline elements (links, marks, etc.). `selection.getNodes()` includes intermediate element nodes, so their `getTextContent()` gets double-counted in the join — e.g. `"Hello worldworld"` vs `"Hello world"`. Tested locally and confirmed the alignment isn't preserved when a link is present.

Details and a suggested fix inline. Also noted how `HeadingNode` handles the same problem via `extractWithChild` for reference.

File: packages/lexical-clipboard/src/clipboard.ts [hunk 1]
```diff
@@ -480,6 +481,40 @@ function exportNodeToJSON<T extends LexicalNode>(node: T): BaseSerializedNode {
...
+    $isRootNode(node) ||
+    node.isInline()
+  ) {
+    return false;
+  }
+  const selectedNodes = selection.getNodes();
+  if (selectedNodes.length === 0) {
+    return false;
+  }
+  const selectedTextContent = selectedNodes
```
Comment (mayrang) [inline packages/lexical-clipboard/src/clipboard.ts:524]: `selection.getNodes()` returns intermediate element nodes too — e.g. for "Hello " + `<a>world</a>`, it returns `[TextNode, LinkNode, TextNode]`. The join then produces `"Hello worldworld"` (LinkNode's `getTextContent()` + child TextNode), which never matches the paragraph's `"Hello world"`. So this function returns `false` for any paragraph containing links, marks, or other inline elements.

Filtering to leaf nodes (e.g. `.filter(n => !$isElementNode(n))`) before the comparison would fix it.

File: packages/lexical-clipboard/src/clipboard.ts [hunk 2]
```diff
@@ -526,6 +561,11 @@ function $appendNodesToJSON(
 
   if (shouldInclude && !shouldExclude) {
     targetArray.push(serializedNode);
+  } else if (
```
Comment (mayrang) [inline packages/lexical-clipboard/src/clipboard.ts:651]: For context — `HeadingNode` already overrides `extractWithChild` to return `true`, so heading alignment is preserved on copy without needing this new function. `ParagraphNode` doesn't override it (defaults to `false`), which is why paragraph formatting gets stripped.

The full-text check makes sense if you only want full-paragraph selections to carry alignment. Just noting the existing mechanism since it covers a related case.

File: packages/lexical/src/__tests__/unit/HTMLCopyAndPaste.test.ts [hunk 3]
```diff
@@ -143,6 +148,43 @@ describe('HTMLCopyAndPaste tests', () => {
...
+          text.select(0, text.getTextContentSize());
+
+          setLexicalClipboardDataTransfer(
+            dataTransfer,
+            $getClipboardDataFromSelection(),
+          );
+
+          target.select(0, target.getChildrenSize());
+
+          const selection = $getSelection();
```
Comment (mayrang) [inline packages/lexical/src/__tests__/unit/HTMLCopyAndPaste.test.ts]: `target.select()` returns `RangeSelection` directly — the `$getSelection()` + `invariant($isRangeSelection)` round-trip can be replaced with `const selection = target.select(0, target.getChildrenSize())`.

**Comment by Rohithmatham12** [comment]
Addressed the review feedback in `752a47ecc`:

- filtered selected intermediate `ElementNode`s out of `$shouldIncludeElementWithSelectedChildren()` so inline elements like links are not double-counted
- added regression coverage for preserving centered paragraph alignment when the selected paragraph contains a link
- simplified the paste target selection to use the `RangeSelection` returned by `target.select()`

Re-ran:

```
node node_modules/vitest/vitest.mjs --project unit packages/lexical/src/__tests__/unit/HTMLCopyAndPaste.test.ts --run
node node_modules/vitest/vitest.mjs --project unit packages/lexical-clipboard/src/__tests__/unit/GetClipboardDataExtension.test.ts packages/lexical-clipboard/src/__tests__/unit/ClipboardImportExtension.test.ts --run
node node_modules/prettier/bin/prettier.cjs --check packages/lexical-clipboard/src/clipboard.ts packages/lexical/src/__tests__/unit/HTMLCopyAndPaste.test.ts
node node_modules/eslint/bin/eslint.js packages/lexical-clipboard/src/clipboard.ts packages/lexical/src/__tests__/unit/HTMLCopyAndPaste.test.ts
```

File: packages/lexical/src/__tests__/unit/HTMLCopyAndPaste.test.ts (hunk 3 above)
Comment (etrepum) [reply packages/lexical/src/__tests__/unit/HTMLCopyAndPaste.test.ts]: `const selection = target.select(0)` is also sufficient, the default for both select arguments is the end (`target.select()` and `target.selectEnd()` are equivalent)

**Comment by etrepum** [comment]
Superseded by #9059
